lib/Backend/BackendUtils.cpp

```diff
@@ -123,6 +131,96 @@ bool isInput(const Placeholder *PH) {
   return false;
 }
 
+bool isOutput(const Placeholder *PH, const IRFunction &F) {
+  auto *weight = F.getWeightForNode(PH);
+  assert(weight && "Weight for a node was not found");
+  for (const auto &use : ValueUses(weight)) {
+    Instruction *user = use.get();
+    // Ignore deallocs.
+    if (isa<DeallocActivationInst>(user)) {
+      continue;
+    }
+    OperandKind kind = use.getOperand().second;
+    if (kind == OperandKind::Out || kind == OperandKind::InOut) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool isInput(const Placeholder *PH, const IRFunction &F) {
+  // Check that the PH is always used as an @in parameter by the current
+  // function.
+  auto *weight = F.getWeightForNode(PH);
+  assert(weight && "Weight for a node was not found");
+  for (const auto &use : ValueUses(weight)) {
+    Instruction *user = use.get();
+    // Ignore deallocs.
+    if (isa<DeallocActivationInst>(user)) {
+      continue;
+    }
+    OperandKind kind = use.getOperand().second;
+    if (kind == OperandKind::In || kind == OperandKind::InOut) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename FUN>
+std::vector<std::tuple<const Placeholder *, bool, bool>>
+getContiguousPlaceHolder(std::vector<const Placeholder *> const &holders,
```

Comment:
I'd suggest to make the type of holders a template type as well. Then you don't need to convert a list into vector in one of the callers.

Reply:
> I'd suggest to make the type of holders a template type as well. Then you don't need to convert a list into vector in one of the callers.

Sure